python/paddle/distributed/passes/auto_parallel_amp.py

```diff
@@ -0,0 +1,715 @@
+            main_block.ops.insert(loss_op_idx + 3, elementwise_mul_grad_op)
+            main_block._sync_with_cpp()
+            elementwise_mul_grad_op = main_block.ops[loss_op_idx + 3]
+            assert elementwise_mul_grad_op.type == "elementwise_mul_grad"
+            naive_set_dist_op_attr_for_program_by_mesh_and_mapping(
+                elementwise_mul_grad_op, ref_mesh, [-1], self.dist_context)
+
+        else:
+            self._scaled_loss = loss
+
+        main_block._sync_with_cpp()
+
+    def _update_loss_scaling(self, grads, found_inf):
+
+        main_block = paddle.static.default_main_program().global_block()
+        main_block._sync_with_cpp()
+
+        check_variable_and_dtype(self._loss_scaling, "prev_loss_scaling",
+                                 ['float32', 'float64'], "update_loss_scaling")
+        check_type(grads, 'x', (tuple, list), 'update_loss_scaling')
+        for e in grads:
+            check_variable_and_dtype(e, "x", ['float16', 'float32', 'float64'],
+                                     'update_loss_scaling')
+            assert self._loss_scaling.dtype == e.dtype, \
+                "The dtype of prev_loss_scaling should be equal to the dtype of x."
+
+        inputs = {
+            'X': grads,
+            'FoundInfinite': found_inf,
+            'PrevLossScaling': self._loss_scaling,
+            'InGoodSteps': self._num_good_steps,
+            'InBadSteps': self._num_bad_steps
+        }
+
+        outputs = {
+            'Out': grads,
+            'LossScaling': self._loss_scaling,
+            'OutGoodSteps': self._num_good_steps,
+            'OutBadSteps': self._num_bad_steps
+        }
+
+        attrs = {
+            'incr_every_n_steps': self.get_attr("incr_every_n_steps"),
+            'decr_every_n_nan_or_inf': self.get_attr("decr_every_n_nan_or_inf"),
+            'incr_ratio': self.get_attr("incr_ratio"),
+            'decr_ratio': self.get_attr("decr_ratio"),
+            'stop_update': self.get_attr("stop_update"),
+            'op_role': OpRole.Backward
+        }
+
+        new_op = main_block.append_op(
+            type='update_loss_scaling',
+            inputs=inputs,
+            outputs=outputs,
+            attrs=attrs)
+
+        new_op_dist_attr = OperatorDistributedAttribute()
+        new_op_dist_attr.process_mesh = global_process_mesh
+        if len(global_process_mesh) > 1:
+            new_op_dist_attr.impl_idx = 0
+        for g in grads:
+            g_dist_attr = self.dist_context.get_tensor_dist_attr_for_program(g)
+            assert g_dist_attr is not None
+            new_op_dist_attr.set_input_dims_mapping(g.name,
+                                                    g_dist_attr.dims_mapping)
+            new_op_dist_attr.set_output_dims_mapping(g.name,
+                                                     g_dist_attr.dims_mapping)
+        self.dist_context.set_op_dist_attr_for_program(new_op, new_op_dist_attr)
+
+        main_block._sync_with_cpp()
```

Comment:
Please add the necessary comments, especially for the methods without underscores.

Reply:
done~